contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,443 @@
+        CouponState calldata _couponState,
+        uint256 _obsFreq,
+        address _autocallSeller
+    ) external initializer {
+        _initialize(_initParams, _vaultParams);
+        _verifyCouponState(
+            _couponState.nCouponType,
+            _couponState.nAB,
+            _couponState.nCB
+        );
+
+        require(_autocallSeller != address(0), "!_autocallSeller");
+        require(_obsFreq > 0 && (period * 1 days) % _obsFreq == 0, "!_obsFreq");
+
+        putOption.nOptionType = _optionType;
+        couponState.nCouponType = _couponState.nCouponType;
+        couponState.nAB = _couponState.AB;
+        couponState.nCB = _couponState.CB;
+
+        nObsFreq = _obsFreq;
+        nPeriod = period;
+        autocallSeller = _autocallSeller;
+        numTotalObs = period * 1 days / _obsFreq;
+    }
+
+    /**
+     * @dev Returns the last observation timestamp and index
+     * @return the last observation timestamp
+     * @return the last observation index
+     */
+    function lastObservation() external view returns (uint256, uint256) {
+        return
+            _lastObservation(
+                IOtoken(optionState.currentOption).expiryTimestamp()
+            );
+    }
+
+    /**
+     * @dev Gets coupons earned
+     * @return the number of coupons earned
+     * @return earned amount in USDC
+     * @return observation timestamp if autocallable, otherwise expiry
+     */
+    function couponsEarned()
+        external
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        (uint256 autocallTS, uint256 nCBBreaches, uint256 lastCBBreach) =
+            _autocallState(
+                IOtoken(optionState.currentOption).expiryTimestamp()
+            );
+
+        (uint256 nCouponsEarned, uint256 earnedAmt, ) =
+            _couponsEarned(nCBBreaches, lastCBBreach);
+
+        return (nCouponsEarned, earnedAmt, autocallTS);
+    }
+
+    /**
+     * @notice Sets next option type
+     * @param _optionType is the next option type
+     */
+    function setOptionType(OptionType _optionType) external onlyOwner {
+        putOption.nOptionType = _optionType;
+        emit OptionTypeSet(_optionType);
+    }
+
+    /**
+     * @notice Sets the new coupon state
+     * @param _couponType is the coupon type
+     * @param _AB is the autocall barrier pct
+     * @param _CB is the coupon barrier pct
+     */
+    function setCouponState(
+        CouponType _couponType,
+        uint256 _AB,
+        uint256 _CB
+    ) external onlyOwner {
+        _verifyCouponState(_couponType, _AB, _CB);
+
+        couponState.nCouponType = _couponType;
+        couponState.nAB = _AB;
+        couponState.nCB = _CB;
+
+        emit CouponStateSet(_couponType, _AB, _CB);
+    }
+
+    /**
+     * @notice Sets the new period and observation period frequency
+     * @param _period is the period
+     * @param _obsFreq is the observation period frequency
```

Comment:
perhaps we could add on the comments that period is in days and obsFreq is a duration in seconds